src/main/java/org/jetbrains/bsp/bazel/server/bsp/resolvers/TargetsResolver.java

```diff
@@ -1,41 +1,174 @@
-    BazelProcessResult bazelProcessResult =
-        bazelRunner
-            .commandBuilder()
-            .query()
-            .withFlag(BazelRunnerFlag.OUTPUT_PROTO)
-            .withTargets(targets)
-            .executeBazelCommand();
+  private Map<String, List<String>> getTargetsOptions(List<String> targets) {
+    BazelProcessResult bazelProcessResult = queryBazel(targets);
+
     Build.QueryResult query = QueryResolver.getQueryResultForProcess(bazelProcessResult);
 
     return query.getTargetList().stream()
         .map(Build.Target::getRule)
-        .collect(
-            Collectors.toMap(
-                Build.Rule::getName,
-                (rule) ->
-                    rule.getAttributeList().stream()
-                        .filter(attr -> attr.getName().equals(compilerOptionsName))
-                        .flatMap(attr -> attr.getStringListValueList().stream())
-                        .collect(Collectors.toList())));
+        .collect(Collectors.toMap(Build.Rule::getName, this::collectRules));
+  }
+
+  private BazelProcessResult queryBazel(List<String> targets) {
+    return bazelRunner
+        .commandBuilder()
+        .query()
+        .withFlag(BazelRunnerFlag.OUTPUT_PROTO)
+        .withTargets(targets)
+        .executeBazelCommand();
+  }
+
+  private List<String> collectRules(Build.Rule rule) {
+    return rule.getAttributeList().stream()
+        .filter(this::isAttributeCompilerOptionsName)
+        .flatMap(attr -> attr.getStringListValueList().stream())
+        .collect(Collectors.toList());
+  }
+
+  private boolean isAttributeCompilerOptionsName(Build.Attribute attribute) {
+    return attribute.getName().equals(compilerOptionsName);
+  }
+
+  private Stream<T> getResultItemForActionGraphParserOptionsTargetsOptionsAndTarget(
+      ActionGraphParser actionGraphParser,
+      Map<String, List<String>> targetsOptions,
+      String target) {
+
+    BuildTargetIdentifier targetIdentifier = new BuildTargetIdentifier(target);
+    List<String> options = targetsOptions.getOrDefault(target, ImmutableList.of());
+    List<String> inputs = actionGraphParser.getInputsAsUri(target, bazelData.getExecRoot());
+
+    return actionGraphParser.getOutputs(target, ACTION_GRAPH_SUFFIXES).stream()
+        .map(this::mapActionGraphOutputsToClassDirectory)
+        .map(
+            classDirectory ->
+                resultItemsCollector.apply(targetIdentifier, options, inputs, classDirectory));
+  }
+
+  private String mapActionGraphOutputsToClassDirectory(String output) {
+    String execPath = Constants.EXEC_ROOT_PREFIX + output;
+
+    return Uri.fromExecPath(execPath, bazelData.getExecRoot()).toString();
+  }
+
+  @FunctionalInterface
+  public interface ResultItemsCollector<T> {
+    T apply(
+        BuildTargetIdentifier target,
+        List<String> options,
+        List<String> classpath,
+        String classDirectory);
+  }
+
+  public static class Builder<T> {
```

Comment:
Would it make sense for `TargetResolver` to live inside this class? To specificy how it pertains to this builder. Or maybe just having them in separate files. Wdyt?

Reply:
I'm not sure I understood the question correctly. `TargetsResolver` should (now it does) have a private constructor and then `Builder` is the only way to create instance of this class

Reply:
But perhaps the `TargetResolver` class code can be inside the builder, that's what I meant

Reply:
No, `TargetResolver` could exist without `Builder` but `Builder` cannot without `TargetResolver`. In general builder pattern looks like this - class + internal builder class

Reply:
It might have been a bit too late for me to have answered, since I meant precisely the opposite of what I wrote :sweat_smile: . The Builder class code being inside the resolver. It's weird for me to see two public classes in a file without this pattern. I'm not against it though, if you disagree, you can go ahead and resolve this and merge :smile:

Reply:
In general I don't like multiple classes in the same file too, but builder pattern is an exception. Maybe we should consider introducing a lombok project, which allows creating builder with one annotation.

Reply:
We can consider doing that in the future, yes